### PR TITLE
Cache and Caching chain have new parameters force and save_cache

### LINKED ISSUE #42: force flag to cache and caching chain

### PR DESCRIPTION
**Pull request by shakukse**
closes #42

### CODE REVIEW

File: flexp/flow/cache.py [hunk 1]
```diff
@@ -367,11 +379,12 @@ def process(self, data):
...
             cache = cache[self.chain_info['chain_hash']]
-            with open(file, 'wb') as f:
-                try:
-                    pickle.dump(cache, f, protocol=4)
-                except:
-                    pickle.dump(cache, f)
+            if self.save_cache:
+                with open(file, 'wb') as f:
+                    try:
+                        pickle.dump(cache, f, protocol=4)
```
Comment (ofilip) [inline flexp/flow/cache.py]: from pickle documentation:
> The optional protocol argument, an integer, tells the pickler to use the given protocol; supported protocols are 0 to HIGHEST_PROTOCOL. If not specified, the default is DEFAULT_PROTOCOL. If a negative number is specified, HIGHEST_PROTOCOL is selected.

File: flexp/flow/caching_chain.py [hunk 2]
```diff
@@ -95,15 +98,17 @@ def process(self, data):
...
-                # key = hashlib.sha256(self.pickle(updated_ids[i])).hexdigest()
-                file = module.get_cache_file_from_id(updated_ids[i])
-                if os.path.exists(file):
-                    log.debug("We skip first {} modules because cache: {} exists (module is {} in a chain)".format(i + 1, file, i + 2))
-                    start = i
-                    break
+        if self.force:
+            log.debug("Force module processing, do not skip anything")
+        else:
+            for i, module in list(enumerate(self.modules))[::-1]:
```
Comment (ofilip) [inline flexp/flow/caching_chain.py:105]: or simpler: reversed(enumerate(self.modules))

**Review by ofilip — APPROVED** [review]
Generally ok, 2 minor suggestions for changes

File: flexp/flow/caching_chain.py (hunk 2 above)
Comment (shakukse) [reply flexp/flow/caching_chain.py:105]: 'enumerate' object is not reversible

File: flexp/flow/cache.py (hunk 1 above)
Comment (shakukse) [reply flexp/flow/cache.py]: agree

**Pull request merged**